x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -249,21 +250,97 @@ void stopDatafeedsIfNecessary(OpenAndClosingIds jobIds, boolean isForce, TimeVal
-                            client,
-                            ClientHelper.ML_ORIGIN,
-                            StopDatafeedAction.INSTANCE,
-                            request,
-                            ActionListener.wrap(r -> listener.onResponse(r.isStopped()), listener::onFailure));
+                        if (isForce) {
+                            // A datafeed with an end time will gracefully close its job when it stops even if it was force stopped.
+                            // If we didn't do anything about this then it would turn requests to force close jobs into normal close
+                            // requests for those datafeeds, which is undesirable - the caller specifically asked for the job to be
+                            // closed forcefully, skipping the final state persistence to save time.  Therefore, before stopping the
+                            // datafeeds in this case we isolate them.  An isolated datafeed will NOT close its associated job under
+                            // any circumstances.  The downside of doing this is that if the stop datafeeds call fails then this API
+                            // will not have closed any jobs, but will have isolated one or more datafeeds, so the failure will have
+                            // an unexpected side effect.  Hopefully the project to combine jobs and datafeeds will be able to improve
+                            // on this.
+                            isolateDatafeeds(jobIds.openJobIds, runningDatafeedIds, ActionListener.wrap(
+                                r -> stopDatafeeds(runningDatafeedIds, true, timeout, listener),
+                                listener::onFailure
+                            ));
+                        } else {
+                            stopDatafeeds(runningDatafeedIds, false, timeout, listener);
+                        }
                     }
                 },
                 listener::onFailure
         ));
     }
 
+    private void stopDatafeeds(List<String> runningDatafeedIds, boolean isForce, TimeValue timeout,
+                               ActionListener<Boolean> listener) {
+        StopDatafeedAction.Request request = new StopDatafeedAction.Request(String.join(",", runningDatafeedIds));
+        request.setForce(isForce);
+        request.setStopTimeout(timeout);
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.ML_ORIGIN,
+            StopDatafeedAction.INSTANCE,
+            request,
+            ActionListener.wrap(
+                r -> listener.onResponse(r.isStopped()),
+                e -> listener.onFailure(ExceptionsHelper.conflictStatusException(
+                    "failed to close jobs as one or more had started datafeeds that could not be stopped: " +
+                        "started datafeeds [{}], error stopping them [{}]", e, request.getDatafeedId(), e.getMessage())
+                )
```

Comment:
It would be awesome to one day have the list of datafeeds that were successfully stopped so that those jobs can continue to close.

one day :D

Reply:
Maybe, but there aren't many reasons why a datafeed would fail to stop to be honest.

What's in this PR is no worse than what we had before, where if you asked to close 100 jobs and one of them had a started datafeed then the whole close request got rejected.  Now we're in the situation where if you ask to close 100 jobs and one of them has a started datafeed and that fails to stop then none of the 100 jobs get closed.

When we eventually reach the stage of the job and datafeed being combined things will naturally get better